### ISSUE TITLE
Exception/Error In Resolvers (Router) can not be catched in the subscription to "activatedRoute.data"

### ISSUE THREAD

**Issue by jbgraug** [angular/angular#13873]
**I'm submitting a ...** (check one with "x")

```
[X ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

I'm using the Router with Resolver interfaces to get some data form a service (backend) before creating a component which will render the data (subscribing to it).
If  an exception occurs in the Resolver the application breaks.  (Error: Uncaught (in promise))

**Expected behavior**

If there is an exception in the Resolver i expect the subscribe error function to be called but the ErrorObservable appears in the "activatedRoute.data.value"  instead.

![image](https://cloud.githubusercontent.com/assets/12108584/21853282/c68edfb2-d816-11e6-8573-9adc8e898020.png)

```
// some component
    ngOnInit() {
        this.activatedRoute.data
            .subscribe(
                (data) => {
                    this.data = data.data] || {};
                    console.log('data was set to ' + this.data)
                },
                (error) => {
                    /// we never get here, no matter there was a Promise.reject(), throw Error(), etc.
                     console.log('ups! Error getting data')
                     alert(error);
                });
    }

// some resolver
resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): BehaviorSubject<any>|Observable<any>|Promise<any>|any {
     return Observable.throw('This always fails');
}
```

**Minimal reproduction of the problem with instructions**

check out the plunker [http://plnkr.co/edit/MaiM6T1COr2ppB2AMANx](http://plnkr.co/edit/MaiM6T1COr2ppB2AMANx)

**What is the motivation / use case for changing the behavior?**

Catch and treat exceptions within the target component
**Please tell us about your environment:**

* **Angular version:** 2.0.X
2.3.1

Yes
* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

All 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.1.1
* **Node (for AoT issues):**`node --version` =

**Comment by DzmitryShylovich** [comment]
Pls add a small repro in plunkr

**Comment by jbgraug** [comment]
Added plunker [http://plnkr.co/edit/MaiM6T1COr2ppB2AMANx](http://plnkr.co/edit/MaiM6T1COr2ppB2AMANx)

**Comment by OscarGalindo** [comment]
Same problem here using 2.4.2. 

I did a "catch" in the resolve which returns an empty observable and check inside component if empty or not.

**Comment by DzmitryShylovich** [comment]
But after an error observable will be closed and you won't get any further updates

**Comment by DzmitryShylovich** [comment]
```
interface ResolvedData {
  data?: any;
  error?: any;
}
resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<ResolvedData> {
    return this.http.get(url).map(r => {data: r.json()}).catch(e => Observable.of({error: e});
  }
```
is more flexible solution as for me, than 1 time observable

**Comment by DzmitryShylovich** [comment]
@jbgraug so do you still want that feature or it can be closed?

**Comment by jbgraug** [comment]
I still want this feature and I think that  it is a must as the Observable is not behaving as it should. 
You need to add extra code as a workaround and if in the future it is fixed it will break the code.

**Comment by needsmatch** [comment]
I agree with @jbgraug I want to be able to catch the resolve service observable error on the ActivateRoute.data observable in my controller.

**Comment by JohannesRudolph** [comment]
@DzmitryShylovich may be able offer more insight here, but I was under the impression that it is the router's job to ensure that a route can only be loaded once all resolvers complete successfully. If you need more involved error handling, you should consider handling errors in the component or use the "workaround" presented by @DzmitryShylovich.

**Comment by OscarGalindo** [comment]
@JohannesRudolph why successfully? i want to load the route once the resolver completes, and an error in a Observable actually completes it. I want the possibility to show an error inside the same component as if it worked

**Comment by JohannesRudolph** [comment]
@OscarGalindo not understanding all the intricacies of rxjs has been one of the biggest headaches I've personally had with angular so far. Observable's can either `complete` or `error`, which are two very different things... 

The router will only navigate if Resolve and CanActivate guards `complete` their returned Observables.

**Comment by OscarGalindo** [comment]
@JohannesRudolph yes, but what i'm trying to say (but my poor english limits me) is that `complete` and `error` terminates the stream, so the Observable is finished in both cases, with an error or a result but it's finished and i think it's fine to have the possibility of catch that error inside the component instead of a resolver

**Comment by jbgraug** [comment]
I totally agree with  @OscarGalindo . 
The problem here is that you can not take over the router error handling (or the resolver's). 
If the the resolver fails, the router simply does not do the navigation. 
So from my modest point of view we have 4 options here:

1. Do nothing. Cons: This is a workaround an it is not intuitive.
2. Let the component decide how to deal with the error. Cons: I don't see any. You always get to the component no matter what. 
3. Import the router in the Resolver and if there is an exception then navigate to whatever url. Cons: I haven't tried it so don't know if it works. Routes are dispersed within the code.
4. Add another attribute 'errorComponent' to the ResolveData type as follows so we can regain control over the error. Cons: You need another component... more boilerplate code.
 `
   {
        path: 'products/:id',
        component: ProductComponent,
        resolve: {
            data: ProductService
            errorComponent: ProductErrorComponent
        },
    }
`
Honestly, I would go for solution 2 which was the initial idea and the most intuitive one.

**Comment by machtyn** [comment]
I used the Resolve example from thoughtram and modified it so that it would break in the route.  
https://plnkr.co/edit/wLosw5
That is,` ContactResolve` will call `ContactsService.failGetContact` which performs `Observable.throw('...');`  When this happens, the exception is logged on the console and the route stays on `ContactsListComponent`.

I want to handle this error at the component level.  But the `ContactsDetailComponent` is never loaded. Should I be handling the error in the `ContactsListComponent` or `ContactsDetailComponent `?  If on CLC, how?

**Comment by samrantmedia** [comment]
Same problem as machtyn. I want to handle a 401 error for example, but it is not possible to handle via the component, without major changes to the app. 

The navbar navigate from the html template using a routerLink attribute. And the component that it is trying to reach, never gets reached because of the 401 in the resolver. 

Therefore the error remains uncaught. There should be a way to catch this error, via the subscribed error function, otherwise what is the point of this error function?

**Comment by gizm0bill** [comment]
Totally agree. Cases like these, if you want to continue after exceptions, should be handled with operators like `retryWhen` /cc @DzmitryShylovich

**Comment by r0b3r70** [comment]
Running into the same issue. The error doesn't reach the component so can't do error handling there.

**Comment by ArtworkAD** [comment]
@samherrmann @r0b3r70 if the error should reach the target component, you can catch the error like this, e.g.:

```
@Injectable()
export class MessageDetailResolverService implements Resolve<ResolveValueEntity<Message>> {

  constructor(private router: Router,
              private messageDatasource: MessageApiDatasource) {}

  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<ResolveValueEntity<Message>> {
    return this.messageDatasource.message(route.paramMap.get('id'))
      .take(1)
      .map(message => {
        if (message) {
          return new ResolveValueEntity(message);
        } else {
          // noinspection JSIgnoredPromiseFromCall
          this.router.navigate(['/404']);
          return null;
        }
      })
      .catch(error => Observable.of(new ResolveValueEntity(null, ErrorMapper.transform(error))));
  }
}
```

and the wrapper

```
/**
 * This class is a generic wrapper for the response that comes from a route resolver.
 * The wrapper allows us to not only return the actual entity T but also an error if
 * the resolve action fails.
 */
export class ResolveValueEntity<T> {
  /**
   * Create a new instance with value of type T.
   * @param {T} value
   * @param {Error} error
   */
  constructor(public value: T,
              public error: Error = null) {}

  /**
   * Check if entity has an error.
   * @returns {boolean}
   */
  hasError(): boolean {
    return this.error !== null;
  }
}
```
This way you can handle the error in the target component. However from a usability perspective it might be better to handle the error in the originating component.

**Comment by Parent5446** [comment]
I'll be completely honest, I thought about this for a while, and did some preliminary research into what would need to be done to fix it, and I'm beginning to disagree that this is a bug, and that @DzmitryShylovich's workaround is actually the proper solution.

The purpose of a resolver is to resolve critical data that *must* exist before the route even loads. If you're dealing with data that does not have to be resolved pre-activation, then you shouldn't use a resolver anyway, and just load the data in ngOnInit(). Therefore, it wouldn't make sense to continue with a successful navigation if a resolver fails with an error, because that would mean that critical data required for the route failed to load. Attempting to catch the error from the resolver is conflating a failure in navigation with a failure in application-specific business logic. If a component expects to successfully load and receive information regarding a failed resolver, then that is still a navigation success, albeit an application failure. Therefore, it is the resolver's responsibility to catch any application errors, and return them as a navigation success in a data structure that properly communicates the state of the resolver to the component.

I think the actual bug here is that ActivatedRoute.data is an Observable in the first place, if it is already resolved when the component is loaded. But this is really just misleading rather than a functional bug.

**Comment by ArtworkAD** [comment]
@Parent5446 agree. If resolve fails, I don't activate the route. Instead I show an error in the source component

    this.router.events
      .takeUntil(this.ngUnsubscribe$)
      .filter(event => event instanceof NavigationError)
      .subscribe(event => {
          // failed to resolve route
      });

**Comment by polyglotinc** [comment]
Since people are still active on this after a year, I'll ask this here...
What officially does the Angular-2/4/5 router do if a resolver throws an error?
_e.g. resolve() returns a promise with a .then() but without .catch()_

**Comment by iagocanalejas** [comment]
Solved this for my case with a `switchMap` operator.

```typescript
import { Injectable } from "@angular/core";
import { Resolve, ActivatedRouteSnapshot, RouterStateSnapshot } from "@angular/router";
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs/Observable";
import { of } from "rxjs/observable/of";
import { tap, catchError, map, filter, switchMap } from "rxjs/operators";


export interface User {
    name: string;
}

@Injectable()
export class TestResolver implements Resolve<User> {

    constructor(private http: HttpClient) {
    }

    public resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<User> {
        return http.get(<url, options>)
        .pipe(switchMap(x => {
            if (x instanceof HttpErrorResponse) {
                return of({name: 'Pepe'});
            }
            return of(x);
        }));
    }
}
```

**Comment by huskerona** [comment]
Has there been any progress on this issue? I have just run into a problem where data loading via ngrx or service in a resolver fails to navigate to the appropriate component. Basically, I can see in ngrx store that the data was loaded, but once resolve finishes (or whatever happens to it), the router does not navigate to the appropriate component. There is no error in the console.

**Comment by NoCopy** [comment]
@Parent5446 
> I'll be completely honest, I thought about this for a while, and did some preliminary research into what would need to be done to fix it, and I'm beginning to disagree that this is a bug, and that @DzmitryShylovich's workaround is actually the proper solution.

I will agree that your logic is sound regarding the reason for a resolver guard in the first place (must resolve data before a route can be resolved) but I disagree that this isn't a fundamental bug for a number of reasons:
1. The type changes depending on result; an error will return a catchable Observable return but a success will result in an Object. You can't subscribe to both as only the error returns the Observable.
2. to @polyglotinc's point there seems to be no official Angular stance on error handling in a resolver, or more importantly, to @KostyaTretyak's point, we don't have a error handler option on the router if the resolver fails; I think you are both right, it shouldn't be an Observable in the first place but we should have a proper error handler.
3. @DzmitryShylovich's is not a proper solution because you don't have any context for the error; for example: did this happen on x component or y component. @DzmitryShylovich's solution only encourages you to handle the error on the resolved component anyways (which goes against to what you're saying about a making sure the data is resolved before you route)

While I realize there are workarounds for all these things it is unnecessarily cumbersome and arguably goes against the ES standards because we are expected to believe that a Promise in = Promise out regardless of level.

Somewhat off topic, but, if anything, in my opinion, the Router in Angular needs a complete re-write. It's easily the worst part.

**Comment by tobias74** [comment]
There are two issues similar to this one, both unresolved as it seems:
https://github.com/angular/angular/issues/15946
https://github.com/angular/angular/issues/25589

**Comment by kosso** [comment]
Is there still no officially documented way to deal with failure in the route resolver?

**Comment by alexandis** [comment]
I can't navivate to the home page from error component after, say, error 403: seems like route is not available and `routerLink` is no use. Do I have the similar case to the described here?

![image](https://user-images.githubusercontent.com/6149843/124512428-04a16c00-dde1-11eb-826f-5df786558518.png)

Any way to resolve it?

**Comment by owl-from-hogvarts** [comment]
My thoughts on this topic are the next.

We should be able to choose failure strategy by our self by defining it in route config. I propose these strategies:
(let assume that failure is when *any* of resolvers return either `error` or `null`)
1. `cancel` - in case of failure *cancel* navigation and stay where we were before navigation
2. `skip` - in case of failure *skip* this route and try to find another. Useful when we navigation from external source to e.g. some non-existent item in list. Then we can just skip the route corresponding to the item and display wild card
3. `fallback` - error terminated observable will be provided to spicified *fallback component* and which should implement some error case logics e.g. display error message and provide some instructions
4. `retry` - this will require `retryWhen` property to route config which will act like `retryWhen` from rxjs and will determine when angular should *retry* to navigate to the route

These variants should cover almost all cases

**Comment by owl-from-hogvarts** [comment]
Also it would be cool to resolver to be able to spicify what strategy should be used in what case, like:
```typescript
class SomeResolver implements Resolve {
  resolve() {
    return {
      failStrategy: "skip",
      payload: somePayload
    }
  }
}
```
Thus if payload will be null, error terminated observable, or rejected promise, router will use strategy defined in returned object.

Cons:
1. it is breaking change since it require to redo resolvers, because now you must put data to `payload` property of returned object but not directly.

I guess there are other ways to implement such behavior which will avoid breaking changes

**Comment by nawlbergs** [comment]
Indeed a resolve on a route should be used for data that is 'required' when your route loads (aka immediately).. 
Therefore, the way I generally handle a navigation failure at a more global level... 
 - Just setup an application message service... 
 - push the error into it (you can attach unique ids if you wanted.. aka `resolveName` or something)
 - have a component at the global level present the issue to the user (or any component listening)

Navigation should not complete on resolve failures/errors...  at least that my opinion.   

```js
constructor(
    // ng
    private router: Router,
    // af lib
    private afApiUserService: AfApiUserService,
    // app
    private appMessageService: AppMessageService,
  ) {

  }

  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<any> {

    const userId = route.params['userId'];

    return this.afApiUserService.GET_BY_ID(userId).pipe(
      mergeMap((response) => {
        const user = response.result;

        if (!user) throw new Error(`User [${userId}] not found!`);

        return of(user);
      }),
      catchError((error) => {
        this.appMessageService.displayError('resolveFailed', { error }); // emit error for any components listening
        return of( { error });
      })
    );
  }
```

**Comment by atscott** [comment]
Errors in guards or resolvers cancel the navigation. These errors result in `NavigationError` event as well as being forwarded to the [error handler](https://angular.dev/api/router/withNavigationErrorHandler#usage-notes). The navigation promise is rejected by default today when this happen but can be configured to resolve `false` like any other failed navigation (https://angular.dev/api/router/RouterConfigOptions#resolveNavigationPromiseOnError). This will likely change to be the default behavior in v20. 

Closing as working as expected.
